miasm2/ir/ssa.py

```diff
@@ -339,3 +339,267 @@ def transform(self, path):
+
+            while todo:
+                block = todo.pop()
+                if block not in frontier:
+                    continue
+
+                # walk through dominance frontier of bloc
+                for node in frontier[block]:
+                    if node in done:
+                        continue
+
+                    # remember blocks which contain phi nodes
+                    if node not in self._phinodes:
+                        self._phinodes[node] = dict()
+
+                    # place empty phi functions for a variable
+                    e = self._gen_empty_phi(variable)
+                    self._phinodes[node][variable] = e.src
+                    done.add(node)
+
+                    if node not in intodo:
+                        intodo.add(node)
+                        todo.add(node)
+
+    @staticmethod
+    def _gen_empty_phi(v):
+        """
+        Generates an empty phi function for a variable
+        :param v: variable
+        :return: ExprAff, empty phi function for v
+        """
+        phi = m2_expr.ExprId("phi", v.size)
+        return m2_expr.ExprAff(v, phi)
+
+    @staticmethod
+    def _fill_phi(*args):
+        """
+        Fills a phi function with variables.
+
+        phi(x.1, x.5, x.6)
+
+        :param args: list of ExprId
+        :return: ExprOp
+        """
+        return m2_expr.ExprOp("phi", *args)
+
+    def _transform_phi_rhs(self, src):
+        """
+        Transforms an expression of a phi function on the
+        right hand side into SSA
+        :param src: expression of a phi function on RHS
+        :return: expression in SSA form
+        """
+        # transform variable on RHS in non-SSA form
+        e = self.reverse_variable(src)
+
+        # transform into SSA form
+        src_ssa = self._transform_expression_rhs(e)
+
+        return src_ssa
+
+    def _rename(self, head):
+        """
+        Transforms each variable expression in the CFG into SSA
+        by traversing the dominator tree in depth-first search.
+
+        1. Transform variables of phi functions on LHS into SSA
+        2. Transform all non-phi expressions into SSA
+        3. Update the successor's phi functions' RHS with current SSA variables
+        4. Save current SSA variable stack for successors in the dominator tree
+
+        Source: Cytron, Ron, et al.
+        "An efficient method of computing static single assignment form"
+        Proceedings of the 16th ACM SIGPLAN-SIGACT symposium on
+        Principles of programming languages (1989), p. 31
+        """
+        # compute dominator tree
+        dominator_tree = self.graph.compute_dominator_tree(head)
+
+        # init SSA variable stack
+        stack = [self._stack_rhs.copy()]
+
+        # walk in DFS over the dominator tree
+        for block in dominator_tree.walk_depth_first_forward(head):
+            # restore SSA variable stack of the predecessor in the dominator tree
+            self._stack_rhs = stack.pop().copy()
```

Comment:
Well, we did this in the beginning. It turned out that is does not work. `stack.append(self._stack_rhs)` does not copy the elements, it stores a reference to `self._stack_rhs`. If the elements within the stack will be modified, this will be applied to the same instance that is located several times in `stack`. 

It may be the case that one of the `.copy()` is unnecessary, but debugging this is hard. Side effects may be hidden in details.